### PR TITLE
chore: check SNI value matches HOST header

### PR DESCRIPTION
**Pull request by stevenpyzhang**
### Description 
Add a new handler to each ServerVerticle (if KSQL rest config enables it) to check the SNI against the Host Header to make sure KSQL is the appropriate server to handle the request.

SNI in Vertx
https://vertx.io/docs/apidocs/io/vertx/core/http/HttpConnection.html#indicatedServerName--

Host Header in Vertx
https://vertx.io/docs/apidocs/io/vertx/core/http/HttpServerRequest.html#host--

### Testing done 
Test on a KSQL server with Confluent UI, saw Kafka requests sent to KSQL and the following logs show up on the KSQL server.
```
[ERROR] 2022-02-05 02:12:53,456 [vert.x-eventloop-thread-0] io.confluent.ksql.api.server.FailureHandler handle - Failed to handle request 421 /2.0/kafka/lkc-qk937/topics
io.confluent.ksql.api.server.KsqlApiException: This request was incorrectly sent to this ksqlDB server
	at io.confluent.ksql.api.server.SniHandler.handle(SniHandler.java:39)
	at io.confluent.ksql.api.server.SniHandler.handle(SniHandler.java:24)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1038)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:137)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:132)
	at io.confluent.ksql.api.server.LoggingHandler.handle(LoggingHandler.java:135)
	at io.confluent.ksql.api.server.LoggingHandler.handle(LoggingHandler.java:35)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1038)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:137)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:132)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:54)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:36)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.http.impl.Http2ServerConnection.onHeadersRead(Http2ServerConnection.java:140)
	at io.netty.handler.codec.http2.Http2FrameListenerDecorator.onHeadersRead(Http2FrameListenerDecorator.java:48)
	at io.netty.handler.codec.http2.Http2EmptyDataFrameListener.onHeadersRead(Http2EmptyDataFrameListener.java:63)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:409)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$1.processFragment(DefaultHttp2FrameReader.java:450)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:457)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:253)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:159)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:173)
	at io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:63)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:378)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:438)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.vertx.core.http.impl.VertxHttp2ConnectionHandler.channelRead(VertxHttp2ConnectionHandler.java:418)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1371)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1234)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1283)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import io.vertx.ext.web.RoutingContext;
+
+public class SniHandler implements Handler<RoutingContext> {
+
+  public SniHandler() {
+  }
+
+  @Override
+  public void handle(final RoutingContext routingContext) {
+    if (routingContext.request().isSSL()) {
```
Comment (alapidas) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java:31]: Does this not need null checks?

Comment (stevenpyzhang) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java:31]: i don't think it's necessary, we don't do it anywhere else in the code. See https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/LoggingHandler.java#L88

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java [hunk 2]
```diff
@@ -114,6 +115,9 @@ private Router setupRouter() {
     final Router router = Router.router(vertx);
 
     router.route().handler(new LoggingHandler(server, loggingRateLimiter));
+    if (server.getConfig().getBoolean(KsqlRestConfig.KSQL_SERVER_SNI_CHECK_ENABLE)) {
```
Comment (stevenpyzhang) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java:118]: In case we need to disable the check.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      final String indicatedServerName = routingContext.request().connection()
+          .indicatedServerName();
+      final String requestHost = routingContext.request().host();
+      if (indicatedServerName != null && requestHost != null) {
+        // sometimes the port is present in the host header, remove it
+        final String requestHostNoPort = requestHost.replaceFirst(":\\d+", "");
+        if (!requestHostNoPort.equals(indicatedServerName)) {
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java:38]: How confident are we that these two are expected to be the same? I wonder if there are other "got'chas" similar to the request host sometimes containing the port. I haven't drilled into the details of where each value comes from myself, am wondering if you got the chance to or not.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+package io.confluent.ksql.api.server;
+
+import static io.netty.handler.codec.http.HttpResponseStatus.MISDIRECTED_REQUEST;
+
+import io.confluent.ksql.rest.Errors;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+public class SniHandler implements Handler<RoutingContext> {
```
Comment (vcrfxia) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java:24]: Add some quick unit tests for edge cases? That this handler only applies for TLS requests, that the port-handling works as expected, etc.

Ideally we'd add an integration test (and also check that the config sets up the new handler in the process) but that might be more trouble than it's worth.

**Review by vcrfxia — COMMENTED** [review]
Thanks @stevenpyzhang -- the approach LGTM but in the absence of automated testing it'd be nice to broaden the types of manual checks for this change. You mentioned you tried using the Confluent UI to issue requests -- what about requests issued from the ksqlDB CLI or via the REST API? It'd be good to check that those work as expected from the handler too.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java (hunk 2 above)
Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java:118]: It's too bad that this flag is a server-only config. It'd be nice if users had a way to override this (at the individual request level) but adding that mechanism for this change is probably overkill.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java (hunk 3 above)
Comment (stevenpyzhang) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java:38]: With the host header, the port is optional so it may be present in some headers https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host

I think with SNI, from various sources I checked, we should only have the hostname present (no port included) https://www.networking4all.com/en/support/faq/what-is-sni

I'm pretty confident that this should be all the gotchas, but let me add a log message if this error is hit to log the SNI and host header. This is also why I made it configurable so we can turn off the check on the server.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java (hunk 4 above)
Comment (stevenpyzhang) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/SniHandler.java:24]: I'll add some unit tests

**Comment by stevenpyzhang** [comment]
>You mentioned you tried using the Confluent UI to issue requests -- what about requests issued from the ksqlDB CLI or via the REST API?

I don't think this is necessary because the only way we hit this is in a browser. Due to something called connection coalescing that browsers do when reusing certs, we may end up with both Kafka /topic and KSQL requests sent to the KSQL server.

In this case, if the KSQL endpoint is hit first, it opens up a connection to the KSQL server. The browser reuses this connection to send the Kafka topic requests to the KSQL server which gives it 404 (after this change it'll be 421).


https://blog.cloudflare.com/http-2-for-web-developers/
https://daniel.haxx.se/blog/2016/08/18/http2-connection-coalescing/

**Review by vcrfxia — COMMENTED** [review]
> I don't think this is necessary because the only way we hit this is in a browser.

Agreed that the motivation for this additional check is only relevant for the browser. I just meant a quick manual test to check that the SNI and request host values for requests sent from the CLI or REST API also match each other. I'd be surprised if they didn't but if they didn't that'd be bad since that means enabling the check would disallow those access patterns.

LGTM except it looks like there's a failing integration test in the build.

**Review by vcrfxia — APPROVED** [review]
LGTM!

**Comment by stevenpyzhang** [comment]
When we `setSni(true)` on the vertx server, there's an unknown issue that occurs when ksqlDB servers have mutual TLS enabled. The `SystemAuthenticationFunctionalTest.MutualAuth` test starts failing with
```
javax.net.ssl.SSLHandshakeException: Empty server certificate chain
	at sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at sun.security.ssl.Alert.createSSLException(Alert.java:117)
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:311)
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:267)
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:258)
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:390)
	at sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:375)
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:377)
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:444)
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:968)
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:955)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:902)
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1548)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1394)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1235)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1284)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

I'm filing a GH issue for https://github.com/confluentinc/ksql/issues/8748 as a follow up. The immediate need for this handler is in an environment where ksqlDB does not have mutual tls enabled so we can have investigate this issue as a follow up.
